packages/pwa-kit-react-sdk/src/ssr/universal/hooks/use-server-effect.tsx

```diff
@@ -0,0 +1,215 @@
+                const data = await boundDidUpdate()
+    
+                return {
+                    [key]: data
+                }
+            })
+        }
+    
+        // To avoid messing with React's hooks order lets make this a
+        // noop for the server.
+        useEffect(wrappedDidUpdate, source)
+    
+        return {
+            loading,
+            error,
+            data
+        }
+    }
+
+    return (...args) => (
+        useServerEffect.apply(this, args)
+    )
+}
+
+/**
+ * Create a provider for use with hooks created with the paired createSeverEffect function.
+ *
+ * @param {string} name - for namspacing and data accessing purposes.
+ * 
+ * @function
+ * 
+ * @returns {ServerEffectProvider}
+ */
+export const createServerEffectContext = (name: string): ServerEffectContext => {
+    console.log('--=-=-=---=-=-=-=---==--==-> createServerEffectContext')
+    const ctxValue = {
+        name,
+        requests: [],
+        data: {},
+        resolved: false
+    } 
+    const Context = React.createContext({})
+    Context.displayName = name
+    const useServerEffect = createServerEffect(Context)
+
+    // This method will trigger and return all the date
+    const resolveEffects: EffectsResolver = async () => {
+        // Early exit for resolved contexts.
+        // if (ctxValue.resolved) {
+        //     console.log('early exit?', ctxValue)
+        //     return {
+        //         [ctxValue.name]: ctxValue.data
+        //     }
+        // }
+
+        const requests = [...ctxValue.requests]
+        ctxValue.requests = []
+        const effectPromises = requests.map((effect) => effect())
+        
+        const effectValues = await Promise.all(effectPromises)
+
+        // Reset the requests
+        ctxValue.requests = []
+
+        // Use this value to stop the hook from recoring events.
+        // ctxValue.resolved = true
+
+        // Turn the data in to a map to se the context value
+        ctxValue.data = effectValues.reduce((acc, curr) => ({
+            ...acc,
+            ...curr
+        }), {})
+
+        return {
+            [name]: {
+                ...ctxValue.data
+            }
+        }
+    }
+
+    // Not the best solution, but we'll use this files global scope to
+    // track all the contexts created so we can resolve all of them 
+    // during the rendering stage.
+    contexts.push(resolveEffects)
+
+    const ServerEffectProvider = ({children, value}) => {
+        if (value) {
+            ctxValue.data = value
+        }
+
+        return (
+            <Context.Provider value={ctxValue}>
+                {children}
+            </Context.Provider>
+        )
+    }
+
+    return {
+        ServerEffectProvider,
+        useServerEffect,
+        resolveEffects
+    }
+}
+
+
+/**
+ * 
+ * @returns 
+ */
+export const resolveAllEffects = async () => {
+    const allData = await Promise.all(contexts.map((resolver) => resolver()))
```

Comment:
```suggestion
    const allData = await Promise.all(contexts.map(async (resolver) => resolver()))
```
If we somehow end up with a sync `resolver` that throws an error, using an `async` method automatically converts that thrown error into a rejected promise. (Not sure if that's something we need to worry about, though.)